{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-01-28 20:30:33.578245: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2024-01-28 20:30:33.598147: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2024-01-28 20:30:33.598166: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2024-01-28 20:30:33.598682: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2024-01-28 20:30:33.602104: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-01-28 20:30:34.031749: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import os\n",
    "\n",
    "from tqdm.notebook import tqdm as tqdm_notebook\n",
    "import keras\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.preprocessing import MultiLabelBinarizer, LabelEncoder\n",
    "\n",
    "import validation.utils as utils\n",
    "     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(106574, 52)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tracks = utils.load('dataset/fma_small/tracks.csv')\n",
    "# features = utils.load('dataset/fma_small/features.csv')\n",
    "# echonest = utils.load('dataset/fma_small/echonest.csv')\n",
    "\n",
    "# np.testing.assert_array_equal(features.index, tracks.index)\n",
    "# assert echonest.index.isin(tracks.index).all()\n",
    "\n",
    "tracks.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8000, 52)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset = tracks.index[tracks['set', 'subset'] <= 'small']\n",
    "assert subset.isin(tracks.index).all()\n",
    "tracks = tracks.loc[subset]\n",
    "tracks.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6400 training examples, 800 validation examples, 800 testing examples\n",
      "Top genres (8): ['Electronic', 'Experimental', 'Folk', 'Hip-Hop', 'Instrumental', 'International', 'Pop', 'Rock']\n",
      "All genres (114): [1, 2, 6, 10, 12, 15, 16, 17, 18, 21, 22, 25, 26, 27, 30, 31, 32, 33, 36, 38, 41, 42, 45, 46, 47, 49, 53, 58, 64, 66, 70, 71, 76, 77, 79, 81, 83, 85, 86, 88, 89, 90, 92, 94, 98, 100, 101, 102, 103, 107, 109, 111, 113, 117, 118, 125, 130, 167, 171, 172, 174, 177, 180, 181, 182, 183, 184, 185, 186, 214, 224, 232, 236, 240, 247, 250, 267, 286, 296, 297, 314, 337, 359, 360, 361, 362, 400, 401, 404, 439, 440, 456, 468, 491, 495, 502, 504, 514, 524, 538, 539, 542, 580, 602, 619, 695, 741, 763, 808, 811, 1032, 1060, 1193, 1235]\n"
     ]
    }
   ],
   "source": [
    "train = tracks.index[tracks['set', 'split'] == 'training']\n",
    "val = tracks.index[tracks['set', 'split'] == 'validation']\n",
    "test = tracks.index[tracks['set', 'split'] == 'test']\n",
    "print('{} training examples, {} validation examples, {} testing examples'.format(*map(len, [train, val, test])))\n",
    "\n",
    "genres = list(LabelEncoder().fit(tracks['track', 'genre_top']).classes_)\n",
    "#genres = list(tracks['track', 'genre_top'].unique())\n",
    "print('Top genres ({}): {}'.format(len(genres), genres))\n",
    "genres = list(MultiLabelBinarizer().fit(tracks['track', 'genres_all']).classes_)\n",
    "print('All genres ({}): {}'.format(len(genres), genres))\n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genre(g: str) -> str:\n",
    "    g = g.lower()\n",
    "    if g == \"rock\" or g == \"instrumental\" or g == \"electronic\": return \"rock\"\n",
    "    return \"pop\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6400,)\n",
      "track_id\n",
      "2         Hip-Hop\n",
      "5         Hip-Hop\n",
      "10            Pop\n",
      "140          Folk\n",
      "141          Folk\n",
      "           ...   \n",
      "154308    Hip-Hop\n",
      "154309    Hip-Hop\n",
      "154413        Pop\n",
      "154414        Pop\n",
      "155066    Hip-Hop\n",
      "Name: (track, genre_top), Length: 8000, dtype: category\n",
      "Categories (16, object): ['Blues', 'Classical', 'Country', 'Easy Listening', ..., 'Pop', 'Rock', 'Soul-RnB', 'Spoken']\n",
      "track_id\n",
      "2         Hip-Hop\n",
      "5         Hip-Hop\n",
      "10            Pop\n",
      "140          Folk\n",
      "141          Folk\n",
      "           ...   \n",
      "153955        Pop\n",
      "153956        Pop\n",
      "154413        Pop\n",
      "154414        Pop\n",
      "155066    Hip-Hop\n",
      "Name: (track, genre_top), Length: 6400, dtype: category\n",
      "Categories (16, object): ['Blues', 'Classical', 'Country', 'Easy Listening', ..., 'Pop', 'Rock', 'Soul-RnB', 'Spoken']\n"
     ]
    }
   ],
   "source": [
    "print(train.shape)\n",
    "labels=tracks[\"track\", \"genre_top\"]\n",
    "print(labels)\n",
    "print(labels[train])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import librosa\n",
    "from NeuralNetwork.CRNN import CRNN_MEL, CRNN_MFCC\n",
    "from NeuralNetwork.CNN import CNN_CHR\n",
    "from NeuralNetwork.combine_model import CombineModel\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"0\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-01-28 20:30:38.567827: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355\n",
      "2024-01-28 20:30:38.591516: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355\n",
      "2024-01-28 20:30:38.591634: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355\n",
      "2024-01-28 20:30:38.592504: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355\n",
      "2024-01-28 20:30:38.592584: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355\n",
      "2024-01-28 20:30:38.592635: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355\n",
      "2024-01-28 20:30:38.644590: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355\n",
      "2024-01-28 20:30:38.644682: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355\n",
      "2024-01-28 20:30:38.644743: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355\n",
      "2024-01-28 20:30:38.644791: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1929] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 21720 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 4090, pci bus id: 0000:01:00.0, compute capability: 8.9\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"sequential\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " conv2d (Conv2D)             (None, 1293, 128, 32)     320       \n",
      "                                                                 \n",
      " max_pooling2d (MaxPooling2  (None, 646, 64, 32)       0         \n",
      " D)                                                              \n",
      "                                                                 \n",
      " conv2d_1 (Conv2D)           (None, 646, 64, 64)       18496     \n",
      "                                                                 \n",
      " max_pooling2d_1 (MaxPoolin  (None, 323, 32, 64)       0         \n",
      " g2D)                                                            \n",
      "                                                                 \n",
      " conv2d_2 (Conv2D)           (None, 323, 32, 128)      73856     \n",
      "                                                                 \n",
      " max_pooling2d_2 (MaxPoolin  (None, 161, 16, 128)      0         \n",
      " g2D)                                                            \n",
      "                                                                 \n",
      " conv2d_3 (Conv2D)           (None, 161, 16, 256)      295168    \n",
      "                                                                 \n",
      " batch_normalization (Batch  (None, 161, 16, 256)      1024      \n",
      " Normalization)                                                  \n",
      "                                                                 \n",
      " time_distributed (TimeDist  (None, 161, 4096)         0         \n",
      " ributed)                                                        \n",
      "                                                                 \n",
      " lstm (LSTM)                 (None, 256)               4457472   \n",
      "                                                                 \n",
      " dense (Dense)               (None, 1024)              263168    \n",
      "                                                                 \n",
      " dense_1 (Dense)             (None, 512)               524800    \n",
      "                                                                 \n",
      " dropout (Dropout)           (None, 512)               0         \n",
      "                                                                 \n",
      " dense_2 (Dense)             (None, 10)                5130      \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 5639434 (21.51 MB)\n",
      "Trainable params: 5638922 (21.51 MB)\n",
      "Non-trainable params: 512 (2.00 KB)\n",
      "_________________________________________________________________\n",
      "None\n",
      "Model: \"sequential_1\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " conv2d_4 (Conv2D)           (None, 1293, 12, 32)      320       \n",
      "                                                                 \n",
      " max_pooling2d_3 (MaxPoolin  (None, 646, 6, 32)        0         \n",
      " g2D)                                                            \n",
      "                                                                 \n",
      " conv2d_5 (Conv2D)           (None, 646, 6, 64)        18496     \n",
      "                                                                 \n",
      " max_pooling2d_4 (MaxPoolin  (None, 323, 3, 64)        0         \n",
      " g2D)                                                            \n",
      "                                                                 \n",
      " conv2d_6 (Conv2D)           (None, 323, 3, 128)       73856     \n",
      "                                                                 \n",
      " max_pooling2d_5 (MaxPoolin  (None, 161, 1, 128)       0         \n",
      " g2D)                                                            \n",
      "                                                                 \n",
      " conv2d_7 (Conv2D)           (None, 161, 1, 256)       295168    \n",
      "                                                                 \n",
      " batch_normalization_1 (Bat  (None, 161, 1, 256)       1024      \n",
      " chNormalization)                                                \n",
      "                                                                 \n",
      " dropout_1 (Dropout)         (None, 161, 1, 256)       0         \n",
      "                                                                 \n",
      " flatten_1 (Flatten)         (None, 41216)             0         \n",
      "                                                                 \n",
      " dense_3 (Dense)             (None, 1024)              42206208  \n",
      "                                                                 \n",
      " dense_4 (Dense)             (None, 512)               524800    \n",
      "                                                                 \n",
      " dropout_2 (Dropout)         (None, 512)               0         \n",
      "                                                                 \n",
      " dense_5 (Dense)             (None, 10)                5130      \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 43125002 (164.51 MB)\n",
      "Trainable params: 43124490 (164.51 MB)\n",
      "Non-trainable params: 512 (2.00 KB)\n",
      "_________________________________________________________________\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/preprocessing/image.py:1451: UserWarning: This ImageDataGenerator specifies `zca_whitening` which overrides setting of`featurewise_std_normalization`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n",
      "Model: \"sequential_2\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " conv2d_8 (Conv2D)           (None, 1293, 20, 32)      320       \n",
      "                                                                 \n",
      " max_pooling2d_6 (MaxPoolin  (None, 646, 10, 32)       0         \n",
      " g2D)                                                            \n",
      "                                                                 \n",
      " conv2d_9 (Conv2D)           (None, 646, 10, 64)       18496     \n",
      "                                                                 \n",
      " max_pooling2d_7 (MaxPoolin  (None, 323, 5, 64)        0         \n",
      " g2D)                                                            \n",
      "                                                                 \n",
      " conv2d_10 (Conv2D)          (None, 323, 5, 128)       73856     \n",
      "                                                                 \n",
      " max_pooling2d_8 (MaxPoolin  (None, 161, 2, 128)       0         \n",
      " g2D)                                                            \n",
      "                                                                 \n",
      " layer_normalization (Layer  (None, 161, 2, 128)       256       \n",
      " Normalization)                                                  \n",
      "                                                                 \n",
      " time_distributed_1 (TimeDi  (None, 161, 256)          0         \n",
      " stributed)                                                      \n",
      "                                                                 \n",
      " lstm_1 (LSTM)               (None, 256)               525312    \n",
      "                                                                 \n",
      " dense_6 (Dense)             (None, 1024)              263168    \n",
      "                                                                 \n",
      " dense_7 (Dense)             (None, 512)               524800    \n",
      "                                                                 \n",
      " dropout_3 (Dropout)         (None, 512)               0         \n",
      "                                                                 \n",
      " dense_8 (Dense)             (None, 10)                5130      \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 1411338 (5.38 MB)\n",
      "Trainable params: 1411338 (5.38 MB)\n",
      "Non-trainable params: 0 (0.00 Byte)\n",
      "_________________________________________________________________\n",
      "None\n",
      "Model: \"sequential_3\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " conv1d (Conv1D)             (None, 30, 32)            128       \n",
      "                                                                 \n",
      " conv1d_1 (Conv1D)           (None, 30, 64)            6208      \n",
      "                                                                 \n",
      " conv1d_2 (Conv1D)           (None, 30, 128)           24704     \n",
      "                                                                 \n",
      " conv1d_3 (Conv1D)           (None, 30, 256)           98560     \n",
      "                                                                 \n",
      " conv1d_4 (Conv1D)           (None, 30, 512)           393728    \n",
      "                                                                 \n",
      " conv1d_5 (Conv1D)           (None, 30, 1024)          1573888   \n",
      "                                                                 \n",
      " conv1d_6 (Conv1D)           (None, 30, 2048)          6293504   \n",
      "                                                                 \n",
      " conv1d_7 (Conv1D)           (None, 30, 4096)          25169920  \n",
      "                                                                 \n",
      " conv1d_8 (Conv1D)           (None, 30, 8192)          100671488 \n",
      "                                                                 \n",
      " flatten_3 (Flatten)         (None, 245760)            0         \n",
      "                                                                 \n",
      " batch_normalization_2 (Bat  (None, 245760)            983040    \n",
      " chNormalization)                                                \n",
      "                                                                 \n",
      " dense_9 (Dense)             (None, 10)                2457610   \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 137672778 (525.18 MB)\n",
      "Trainable params: 137181258 (523.30 MB)\n",
      "Non-trainable params: 491520 (1.88 MB)\n",
      "_________________________________________________________________\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "#new\n",
    "model_mel = CRNN_MEL((1293,128,1))\n",
    "model_chr = CNN_CHR((1293,12,1))\n",
    "model_mfcc = CRNN_MFCC((1293, 20, 1))\n",
    "\n",
    "# model_mel = CRNN_MEL((128,293,1))\n",
    "# model_chr = CNN_CHR((12,293,1))\n",
    "# model_mfcc = CRNN_MFCC(( 20,1293, 1))\n",
    "combine_model = CombineModel(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_mel.load_model(\"./model/model_mel.h5\")\n",
    "model_chr.load_model(\"./model/model_chr.h5\")\n",
    "model_mfcc.load_model(\"./model/model_mfcc.h5\")\n",
    "combine_model.model = keras.models.load_model(\"./model/combine.h5\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "aca2b256d21b42bd8c8f8f4b60923535",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/800 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(128, 1293)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-01-28 20:30:47.468663: I external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:454] Loaded cuDNN version 8907\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/1 [==============================] - 0s 479ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1291)\n",
      "1/1 [==============================] - 0s 27ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1291)\n",
      "1/1 [==============================] - 0s 14ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1293)\n",
      "1/1 [==============================] - 0s 12ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1291)\n",
      "1/1 [==============================] - 0s 18ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1291)\n",
      "1/1 [==============================] - 0s 24ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1291)\n",
      "1/1 [==============================] - 0s 28ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1291)\n",
      "1/1 [==============================] - 0s 23ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1293)\n",
      "1/1 [==============================] - 0s 16ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1293)\n",
      "1/1 [==============================] - 0s 27ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1293)\n",
      "1/1 [==============================] - 0s 17ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1293)\n",
      "1/1 [==============================] - 0s 21ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1291)\n",
      "1/1 [==============================] - 0s 13ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1293)\n",
      "1/1 [==============================] - 0s 14ms/step\n",
      "in user code:\n",
      "\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2440, in predict_function  *\n",
      "        return step_function(self, iterator)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2425, in step_function  **\n",
      "        outputs = model.distribute_strategy.run(run_step, args=(data,))\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2413, in run_step  **\n",
      "        outputs = model.predict_step(data)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/training.py\", line 2381, in predict_step\n",
      "        return self(x, training=False)\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py\", line 70, in error_handler\n",
      "        raise e.with_traceback(filtered_tb) from None\n",
      "    File \"/home/lumiani/miniconda3/envs/cs182/lib/python3.11/site-packages/keras/src/engine/input_spec.py\", line 298, in assert_input_compatibility\n",
      "        raise ValueError(\n",
      "\n",
      "    ValueError: Input 0 of layer \"sequential_1\" is incompatible with the layer: expected shape=(None, 12, 1293, 1), found shape=(None, 1293, 12, 1)\n",
      "\n",
      "(128, 1293)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 20\u001b[0m\n\u001b[1;32m     16\u001b[0m     right \u001b[38;5;241m=\u001b[39m gap \u001b[38;5;241m-\u001b[39m left\n\u001b[1;32m     17\u001b[0m     mel_pic \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mhstack((np\u001b[38;5;241m.\u001b[39mhstack((np\u001b[38;5;241m.\u001b[39mzeros((\u001b[38;5;241m128\u001b[39m, left)), mel_pic)), np\u001b[38;5;241m.\u001b[39mzeros((\u001b[38;5;241m128\u001b[39m, right))))\n\u001b[0;32m---> 20\u001b[0m harmonic \u001b[38;5;241m=\u001b[39m \u001b[43mlibrosa\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meffects\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mharmonic\u001b[49m\u001b[43m(\u001b[49m\u001b[43my\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m##\u001b[39;00m\n\u001b[1;32m     21\u001b[0m harmonic_features \u001b[38;5;241m=\u001b[39m librosa\u001b[38;5;241m.\u001b[39mfeature\u001b[38;5;241m.\u001b[39mchroma_cqt(y\u001b[38;5;241m=\u001b[39mharmonic, sr\u001b[38;5;241m=\u001b[39msr)  \u001b[38;5;66;03m##\u001b[39;00m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m harmonic_features\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m] \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m1293\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/cs182/lib/python3.11/site-packages/librosa/effects.py:142\u001b[0m, in \u001b[0;36mharmonic\u001b[0;34m(y, **kwargs)\u001b[0m\n\u001b[1;32m    139\u001b[0m stft \u001b[38;5;241m=\u001b[39m core\u001b[38;5;241m.\u001b[39mstft(y)\n\u001b[1;32m    141\u001b[0m \u001b[38;5;66;03m# Remove percussives\u001b[39;00m\n\u001b[0;32m--> 142\u001b[0m stft_harm \u001b[38;5;241m=\u001b[39m \u001b[43mdecompose\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhpss\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstft\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m    144\u001b[0m \u001b[38;5;66;03m# Invert the STFTs\u001b[39;00m\n\u001b[1;32m    145\u001b[0m y_harm \u001b[38;5;241m=\u001b[39m core\u001b[38;5;241m.\u001b[39mistft(stft_harm, dtype\u001b[38;5;241m=\u001b[39my\u001b[38;5;241m.\u001b[39mdtype, length\u001b[38;5;241m=\u001b[39my\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m])\n",
      "File \u001b[0;32m~/miniconda3/envs/cs182/lib/python3.11/site-packages/librosa/decompose.py:393\u001b[0m, in \u001b[0;36mhpss\u001b[0;34m(S, kernel_size, power, mask, margin)\u001b[0m\n\u001b[1;32m    390\u001b[0m harm[:] \u001b[38;5;241m=\u001b[39m median_filter(S, size\u001b[38;5;241m=\u001b[39mharm_shape, mode\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreflect\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    392\u001b[0m perc \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mempty_like(S)\n\u001b[0;32m--> 393\u001b[0m perc[:] \u001b[38;5;241m=\u001b[39m \u001b[43mmedian_filter\u001b[49m\u001b[43m(\u001b[49m\u001b[43mS\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msize\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mperc_shape\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mreflect\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    395\u001b[0m split_zeros \u001b[38;5;241m=\u001b[39m margin_harm \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m margin_perc \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    397\u001b[0m mask_harm \u001b[38;5;241m=\u001b[39m util\u001b[38;5;241m.\u001b[39msoftmask(\n\u001b[1;32m    398\u001b[0m     harm, perc \u001b[38;5;241m*\u001b[39m margin_harm, power\u001b[38;5;241m=\u001b[39mpower, split_zeros\u001b[38;5;241m=\u001b[39msplit_zeros\n\u001b[1;32m    399\u001b[0m )\n",
      "File \u001b[0;32m~/miniconda3/envs/cs182/lib/python3.11/site-packages/scipy/ndimage/_filters.py:1588\u001b[0m, in \u001b[0;36mmedian_filter\u001b[0;34m(input, size, footprint, output, mode, cval, origin, axes)\u001b[0m\n\u001b[1;32m   1541\u001b[0m \u001b[38;5;129m@_ni_docstrings\u001b[39m\u001b[38;5;241m.\u001b[39mdocfiller\n\u001b[1;32m   1542\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmedian_filter\u001b[39m(\u001b[38;5;28minput\u001b[39m, size\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, footprint\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, output\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   1543\u001b[0m                   mode\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreflect\u001b[39m\u001b[38;5;124m\"\u001b[39m, cval\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.0\u001b[39m, origin\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m*\u001b[39m, axes\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m   1544\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1545\u001b[0m \u001b[38;5;124;03m    Calculate a multidimensional median filter.\u001b[39;00m\n\u001b[1;32m   1546\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1586\u001b[0m \u001b[38;5;124;03m    >>> plt.show()\u001b[39;00m\n\u001b[1;32m   1587\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1588\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_rank_filter\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msize\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfootprint\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moutput\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcval\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1589\u001b[0m \u001b[43m                        \u001b[49m\u001b[43morigin\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmedian\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxes\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/cs182/lib/python3.11/site-packages/scipy/ndimage/_filters.py:1489\u001b[0m, in \u001b[0;36m_rank_filter\u001b[0;34m(input, rank, size, footprint, output, mode, cval, origin, operation, axes)\u001b[0m\n\u001b[1;32m   1485\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[1;32m   1486\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mA sequence of modes is not supported by non-separable rank \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1487\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfilters\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1488\u001b[0m mode \u001b[38;5;241m=\u001b[39m _ni_support\u001b[38;5;241m.\u001b[39m_extend_mode_to_code(mode)\n\u001b[0;32m-> 1489\u001b[0m \u001b[43m_nd_image\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrank_filter\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrank\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfootprint\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moutput\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcval\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1490\u001b[0m \u001b[43m                      \u001b[49m\u001b[43morigins\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1491\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m temp_needed:\n\u001b[1;32m   1492\u001b[0m     temp[\u001b[38;5;241m.\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;241m.\u001b[39m] \u001b[38;5;241m=\u001b[39m output\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "acc_cnt = 0\n",
    "all_cnt = 0\n",
    "for i in tqdm_notebook(test):\n",
    "    if labels[i] in ['Experimental', 'Instrumental', 'International']: continue\n",
    "    filename = f\"{i:06d}.mp3\"\n",
    "    file_path = os.path.join(\"/home/lumiani/Projects/IML/CS182-FinalProject/dataset/fma_small\", filename[:3], filename)\n",
    "    if not os.path.exists(file_path): continue\n",
    "    y, sr = librosa.load(file_path)\n",
    "    spectrogram = librosa.feature.melspectrogram(y=y, sr=sr)\n",
    "\n",
    "    mel_pic = librosa.power_to_db(spectrogram, ref=np.max)\n",
    "    print(mel_pic.shape)\n",
    "    if mel_pic.shape[1] < 1293:\n",
    "        gap = 1293 - mel_pic.shape[1]\n",
    "        left = gap // 2\n",
    "        right = gap - left\n",
    "        mel_pic = np.hstack((np.hstack((np.zeros((128, left)), mel_pic)), np.zeros((128, right))))\n",
    "\n",
    "\n",
    "    harmonic = librosa.effects.harmonic(y)  ##\n",
    "    harmonic_features = librosa.feature.chroma_cqt(y=harmonic, sr=sr)  ##\n",
    "\n",
    "    if harmonic_features.shape[1] < 1293:\n",
    "        gap = 1293 - harmonic_features.shape[1]\n",
    "        left = gap // 2\n",
    "        right = gap - left\n",
    "        harmonic_features = np.hstack((np.hstack((np.zeros((12, left)), harmonic_features)), np.zeros((12, right))))\n",
    "\n",
    "\n",
    "    mfccs = librosa.feature.mfcc(y=y, sr=sr)\n",
    "\n",
    "    if mfccs.shape[1] < 1293:\n",
    "        gap = 1293 - mfccs.shape[1]\n",
    "        left = gap // 2\n",
    "        right = gap - left\n",
    "        mfccs = np.hstack((np.hstack((np.zeros((20, left)), mfccs)), np.zeros((20, right))))\n",
    "\n",
    "\n",
    "    duration = len(y) / sr\n",
    "\n",
    "    # Assuming model1, model2, model3 are already defined with different input shapes but the same output shape\n",
    "    try:\n",
    "        _, pre_mel = model_mel.predict(mel_pic, duration)\n",
    "        _, pre_chr = model_chr.predict(harmonic_features, duration)\n",
    "        _, pre_mfcc = model_mfcc.predict(mfccs, duration)\n",
    "\n",
    "        pre_mel=pre_mel.flatten()\n",
    "        pre_chr=pre_chr.flatten()\n",
    "        pre_mfcc=pre_mfcc.flatten()\n",
    "\n",
    "        merge_array=[]\n",
    "        for arr in [pre_mel, pre_chr, pre_mfcc]:\n",
    "            merge_array.extend(arr)\n",
    "        merge_array=np.expand_dims(merge_array, axis=0)\n",
    "        pre = combine_model.model.predict(merge_array)\n",
    "        class_idx = np.argmax(pre, axis=1)[0]\n",
    "        acc_cnt += ([\"pop\", \"classical\", \"pop\", \"rock\", \"pop\", \"jazz\", \"rock\", \"pop\", \"pop\", \"rock\"][class_idx] == genre(labels[i]))\n",
    "        all_cnt += 1\n",
    "    except Exception as e:\n",
    "        print(e)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "ZeroDivisionError",
     "evalue": "division by zero",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[33], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtest accuracy: \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\u001b[43macc_cnt\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m/\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mall_cnt\u001b[49m))\n",
      "\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"
     ]
    }
   ],
   "source": [
    "print(\"test accuracy: {}\".format(acc_cnt / all_cnt))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cs182",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
